feature-libs/cart/quick-order/core/services/quick-order.service.ts

```diff
@@ -152,6 +168,92 @@ export class QuickOrderService implements QuickOrderFacade {
     );
   }
 
+  /**
+   * Add deleted entry
+   */
+  addDeletedEntry(entry: OrderEntry, clearTimeout: boolean = true): void {
+    const deletedEntries = this.deletedEntries$.getValue() || {};
+    const productCode = entry.product?.code;
+
+    if (productCode) {
+      deletedEntries[productCode] = entry;
+
+      this.deletedEntries$.next(deletedEntries);
+
+      if (clearTimeout) {
+        const subscription: Subscription = timer(
+          this.deletionClearTimeout
+        ).subscribe(() => {
+          if (productCode) {
+            this.clearDeletedEntry(productCode);
+          }
+        });
+        this.clearMessageTimeouts[productCode] = subscription;
+      }
+    }
+  }
+
+  /**
+   * Return deleted entries
+   */
+  getDeletedEntries(): Observable<DeletedEntriesObject> {
+    return this.deletedEntries$;
+  }
+
+  /**
+   * Undo deleted entry
+   */
+  undoDeletedEntry(productCode: string): void {
+    this.cleanUpDeletedEntry(productCode, true);
+  }
+
+  /**
+   * Clear deleted entry from the list
+   */
+  clearDeletedEntry(productCode: string): void {
+    this.cleanUpDeletedEntry(productCode);
+  }
+
+  /**
+   * Clear all deleted entry timeout subscriptions
+   */
+  clearTimeoutSubscriptions(): void {
+    Object.values(this.clearMessageTimeouts).forEach(
+      (subscription: Subscription) => subscription.unsubscribe()
+    );
+  }
+
+  /**
+   * Remove deleted entry and optionally add it back to entries
+   */
+  protected cleanUpDeletedEntry(
+    productCode: string,
+    addEntry: boolean = false
+  ): void {
```

Comment:
I believe that this function should not be responsible for adding it back to entries list. That should be done in undoDeletedEntry method